---
title: "Deseq2 Analysis and R Shiny Application"
author: "Suneeta Modekurty"
date: "01/01/2021"
output: pdf_document

Data:
A matrix of gene counts generated by Salmon: Changed for convenience to GeneCounts.csv
A csv file containing phenotype information: Changed for convenience to Pheno.csv

Problem: To perform differential expression analysis while controlling for age and gender with DESEQ2. Using the count matrix and phenotype file, use DESEQ2 to determine the differential expression, controlling for gender and age.  

# Please run the .rmd to get expected results.
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, message=FALSE}
tinytex::install_tinytex()
```

```{r, echo=FALSE, message=FALSE}
library(DESeq2) ####Load All Pakages
library(ggplot2)
library(RColorBrewer)
library(pheatmap)
library(BiocParallel)
library(tximportData)
library(ggrepel)
library(tidyverse)
```



```{r}
####################
samples <- read.csv("pheno.csv", header=TRUE) ## Read metadata File
head(samples)
#View(samples)
dim(samples)

######################
data <- read.csv("GeneCounts.csv",header=TRUE,sep=",") ####Read the data file
#View(data)
head(data)
dim(data)


```


```{r, message=FALSE, warning=FALSE}
####################  
dds <- DESeqDataSetFromMatrix(countData =  data, colData = samples, design = ~ Diagnosis, tidy = T) ###make the dds(dds is now ready for DESeq)
######################
dds <- DESeq(dds)	
resultsNames(dds)
res <- results(dds)
res
```



```{r, message=FALSE}
#contrast <- c("Pathologic Aging","Control")
#dds$Diagnosis <- factor(dds$Diagnosis, levels = c("Pathologic.Aging","Control"))

res <- results(dds, contrast=c("Diagnosis","Pathologic.Aging","Control"))
#res <- results(dds, contrast)
resOrdered <- res[order(res$pvalue),]
summary(res)
sum(res$padj < 0.1, na.rm=TRUE)

write.csv(res, file = "data.csv") 
```


```{r, message=FALSE}
rld <- rlog(dds, blind=FALSE)
write.csv(assay(rld), file = "expression.csv")
ntd <- normTransform(dds)
mat = assay(rld)[ head(order(res$padj),10), ] 	
mat = mat - rowMeans(mat)		 
df = as.data.frame(colData(rld)[,c("Diagnosis")])
colnames(df) = "Diagnosis" 
rownames(df) = colnames(mat) 
png(file ="Diagnosis.png", width = 10, height = 10)
pheatmap(mat, annotation_col=df)		 
#dev.off()	 
```


#res <- results(dds)
mcols(res, use.names=TRUE)
summary(res)
pdf(file ="MA.pdf", width = 10, height = 10)
plotMA(res, ylim=c(-7,7))
dev.off()

```{r}
plotMA(res, ylim=c(-7,7))
#plotMA(resLFC, ylim=c(-2,2))
```

# Controling with age and gender 
```{r}
colData(dds)
ddsMF <- dds
ddsMF$Sex<- as.factor(ddsMF$Sex)
ddsMF$AgeAtDeath<- as.factor(ddsMF$AgeAtDeath)
levels(ddsMF$Sex)
levels(ddsMF$AgeAtDeath)
```


```{r, message=FALSE}
design(ddsMF) <- formula(~ Sex + AgeAtDeath + Diagnosis+ Sex:Diagnosis)
ddsMF <- DESeq(ddsMF)

resultsNames(ddsMF)

resMF <- results(ddsMF)
head(resMF)
```


```{r}
res <- results(ddsMF, contrast=c("Diagnosis","Pathologic.Aging","Control"))
plotMA(res, ylim=c(-7,7))
```


```{r}
resOrdered <- res[order(res$pvalue),]
summary(res)
sum(res$padj < 0.1, na.rm=TRUE)

write.csv(res, file = "data1.csv")
```


```{r}
rld <- rlog(ddsMF, blind=FALSE)
assay(rld)
ntd <- normTransform(ddsMF)
mat = assay(rld)[ head(order(res$padj),10), ] 	
mat = mat - rowMeans(mat)		 
df = as.data.frame(colData(rld)[,c("Diagnosis")])
colnames(df) = "Diagnosis" 
rownames(df) = colnames(mat) 
#pdf(file ="Diagnosis.pdf", width = 10, height = 10)
pheatmap(mat, annotation_col=df)		 
#dev.off()	 
```

```{r}
results = as.data.frame(dplyr::mutate(as.data.frame(res), sig=ifelse(res$pvalue<0.05, "Fold changes", "Significant&Fold change")), row.names=rownames(res))
head(results)
DEgenes_DESeq <- results[which(abs(results$log2FoldChange) > log2(2) & results$pvalue<0.05),]
write.csv(DEgenes_DESeq, file = "results2.csv")
```

# Controling with variable gender only

```{r, message=FALSE}
design(ddsMF) <- formula(~ Sex + Diagnosis + Sex:Diagnosis)
ddsMF1 <- DESeq(ddsMF)
resultsNames(ddsMF1)
resMF1 <- results(ddsMF1)
head(resMF1)
```


```{r}
res1 <- results(ddsMF1, contrast=c("Diagnosis","Pathologic.Aging","Control"))
plotMA(res1, ylim=c(-7,7))
```
```{r}


# Presentation of output: Present the results in an R Shiny application
# a.	An ordered table with the results.
# used this link https://rstudio.github.io/DT/shiny.html 

# b.	A way to let the user pick any gene and have the system present a chart with the expression of cases and controls. This is a graph  which will use an intermediate table which is normalized 

# intermediate files =  rld (expression.csv)
# I took av of all case samples in a col
# I took av of all control samples in a col
# Got an input table that will have a column of geneIDs, a col of av of cases and a col of av of controls.


# A box in shiny where a geneID can be typed, Using that data, in P.aging,and in Controls,  it  shows exp.value in boxplots by entering submit.

# Find out more about building applications with Shiny here:
#
#    http://shiny.rstudio.com/
#
##library(DT)
library(data.table)
library(dplyr)
#library(ggplot2)
library(shiny)
```


```{r}
ui <- shinyUI(fluidPage(
    fluidRow(
        column(
            6, 
            h1("Gene Expression  boxplots"),
            p("Case(P.A) Vs. Control"),
            DT::dataTableOutput("table_results2")
        ),
        column(1),
        column(
            5,
            h1("Pick the GeneID"),
            textInput(inputId = "gene", 
                      label = "Type GeneID in the box", 
                      value = "ABCA13"),
            plotOutput("selected_gene", height = "450px")
        )
    )
))

```

```{r}
results2 <- read.csv("results2.csv", row.names = "X") %>%
    select(log2FoldChange, padj)
pheno <- read.csv("Pheno.csv")
case <- pheno %>%
    filter(Diagnosis == "Pathologic Aging") %>%
    pull(UID)
control <- pheno %>%
    filter(Diagnosis == "Control") %>%
    pull(UID)

expression <- fread("expression.csv")

server <- shinyServer(function(input, output) {
    output$table_results2 <- DT::renderDataTable({
        results2
    },
    rownames = TRUE,
    filter = list(position = 'top', clear = FALSE),
    options = list(
        pageLength = 10,
        processing=FALSE
    ))
    
    output$selected_gene <- renderPlot({
        gene <- input$gene
        gene_expr <- expression[V1 == gene] %>%
            .[, V1 := NULL]
        # remove "X" at the beginning of column name, which is add to column
        # name as number is not allowed to start
        gene_expr <- data.table(UID = sub(".", "", names(gene_expr)), 
                                expr = as.matrix(gene_expr)[1,]) %>%
            .[UID %in% case, diagnosis := "Pathologic Aging"] %>%
            .[UID %in% control, diagnosis := "Control"]
        
        ggplot(gene_expr, aes(diagnosis, expr)) +
            geom_jitter(width = 0.1, height = 0, color = "grey") +
            geom_boxplot(fill = NA) +
            labs(x = "Diagnosis", 
                 y = "Expression") +
            theme_bw()
        
    })
})
```

```{r}
# Run the application 
shinyApp(ui = ui, server = server)
```